### PR TITLE
DATA-891 Allow editing of URIs by sysadmins

### PR DESCRIPTION
**Pull request by markstuart**
Also move to UUID1 generation to prevent issues with renaming causing possible URI uniqueness clashes.

### CODE REVIEW

**Comment by markstuart** [comment]
Screenshot from 2023-04-21 08-42-49
Screenshot from 2023-04-21 08-26-50

**Comment by markstuart** [comment]
Oops, I just realised that I didn't meet the requirements agreed upon here... regenerating the URI is supposed to retain the old URI and archive it rather than edit the existing one :face_with_spiral_eyes:

```
URI's need to be able to be regenerated if the remit of the organisation changes

This will keep the old entity record and mark it as archived, so that the future resolver is able to track from the original URI to the current one if needed

The Entity Name of a URI should be able to be updated without generating a new URI if the organisation name changes but its remit stays the same.

Suggested that we have an edit view with information about when you would regenerate vs change the name.

Add the date of creation of the URIs to the list view, and a link the edit form for each URI.

Move to using a strategy of a timestamp based identifier generation, so that URI/identifier clashes cannot occur even after renaming etc
```

File: ckanext/dia/views.py [hunk 1]
```diff
@@ -75,3 +81,43 @@ def list():
...
+
+    data = {'type': current_uri.type, 'name': current_uri.name}
+    vars = {'data': data, 'errors': None, 'error_summary': None, 'update': True}
+    if request.method == 'POST':
+        data_dict = dict(request.form)
+        data_dict['type'] = current_uri.type
+        # checkbox field name is present in request.form if
+        # checked, missing if not checked
+        if 'regenerate' in data_dict:
+            data_dict['regenerate'] = True
```
Comment (niallmcandrew-work) [inline ckanext/dia/views.py:101]: Explicitly set to true or false this way.

```suggestion
        data_dict['regenerate'] = 'regenerate' in data_dict
```

**Review by niallmcandrew-work — APPROVED** [review]
Makes sense to me, only one minor comment.

File: ckanext/dia/model.py [hunk 2]
```diff
@@ -66,10 +66,12 @@ def define_table():
 
 def name_and_type_unique(name, context):
     type = context.get('type')
+    model_id = context.get('id', 0)
     result = MintedURI.Session.query(MintedURI)\
         .filter(func.lower(MintedURI.name) == name.lower())\
         .filter(func.lower(MintedURI.type) == type.lower())\
         .filter(MintedURI.superseded_by == None)\
+        .filter(MintedURI.id != int(model_id))\
```
Comment (markstuart) [inline ckanext/dia/model.py:74]: This means that when editing an existing instance, we only validate against other instances with the same name and type, not the one being edited.

File: ckanext/dia/templates/uris/edit.html [hunk 3]
```diff
@@ -54,7 +43,18 @@ <h2 class="module-heading"><i class="fa fa-info-circle"></i> {{ _('URIs for use
...
+            {{ form.info(_('Only tick this box if the entity that this URI represents is changing its remit or purpose.'), classes=['info-help-tight', 'small', 'bold']) }}
+            {{ form.checkbox('regenerate',
+                label=_("Regenerate URI?"),
+                checked=data.regenerate,
+                classes=[""])
+            }}
+
         </fieldset>
+        <div class="form-actions">
+            <a class="btn" href="/uri/">{{_('Cancel')}}</a>
```
Comment (markstuart) [inline ckanext/dia/templates/uris/edit.html:56]: Kervin requested a change to the info text, moving the regen checkbox lower, and a cancel link

File: ckanext/dia/templates/uris/success.html [hunk 4]
```diff
@@ -21,19 +21,19 @@ <h2 class="module-heading"><i class="fa fa-info-circle"></i> {{ _('URIs for use
 {% endblock %}
 
 {% block primary_content %}
-    <form method="POST">
```
Comment (markstuart) [inline ckanext/dia/templates/uris/success.html:24]: I probably copypasta'd this in here, there was no reason for this to be a form.

File: ckanext/dia/templates/uris/success.html [hunk 5]
```diff
@@ -21,19 +21,19 @@ <h2 class="module-heading"><i class="fa fa-info-circle"></i> {{ _('URIs for use
...
+        {% endif %}
+    <br /><strong>{{data.uri}}</strong></p>
 
-            <p>{{_('This has been reserved against the name')}} {{data.name}}</p>
-        </fieldset>
-    </form>
+    <p>{{_('This has been reserved against the name')}} "{{data.name}}"</p>
+
+    <div class="form-actions">
+        <a class="btn btn-primary" href="/uri/">{{_('Back to URI listing')}}</a>
```
Comment (markstuart) [inline ckanext/dia/templates/uris/success.html:37]: Added this back button here to make it clearer what to do after success.

**Review by niallmcandrew-work — APPROVED** [review]

**Pull request merged**